Approved/3-multidimensional-arrays.md

```diff
@@ -0,0 +1,701 @@
+array
+    w/ Strides <- [array::Strides[0], 2 * array::Strides[1]]
+    w/ Shape <- [array::Shape[0], array::Shape[1] / 2]
+```
+
+Reversing an axis can be implemented by using negative strides and modifications to `::Offset`.
+
+By implementing multidimensional arrays in this way, we can reuse the existing infrastructure for immutable single-dimensional arrays.
+Moreover, in many cases, multiple distinct multidimensional arrays can share the same `::Data` item without requiring a copy.
+For example, in the variable binding `let view = array[0..2..., 0..2...];`, `view::Data` and `array::Data` can be the same single-dimensional array, as the difference between `view` and `data` can be expressed by only modifying `::Strides` and `::Shape`.
+The cases where copies may still be required are when reshaping from more indices to less, when using `w/` to update slices, or if a particular intrinsic function or operation is implemented in simulation by interoperating with native libraries such as BLAS and LAPACK.
+
+## Timeline
+
+This proposal does not depend on any other proposals (though it can be made easier-to-use in combination with other proposals; see _Anticipated Interactions with Future Modifications_, below).
+
+Required implementation steps:
+
+- Adapt compiler to recognize new types, subscript expressions, copy-and-replace expressions, and new literal expressions.
+- Implement new data structures in QIR and simulation runtime to represent values of new type, and fundamental operations on new type (e.g.: `w/` expressions).
+- Design, review, and approve API design for extensions to the `Microsoft.Quantum.Arrays namespace` to support the new feature.
+- Implement new library functions from previous step.
+- Document new multidimensional arrays in language specification and conceptual documentation.
+
+# Further Considerations
+
+## Related Mechanisms
+
+This proposal generalizes the existing array feature in Q#.
+As such, the new features introduced by this proposal are designed to keep to user expectations formed by existing features.
+In particular:
+
+- Like existing arrays, multidimensional arrays are immutable values.
+- Mutability can be handled using `set ... w/= ...;` statements in the same fashion as existing 1D arrays.
+- Multidimensional arrays can be used as collections in loops.
+- There are no subtyping relationships between array types. In particular, `[|'T|]` is not a subtype of `'T[][]` but a distinct type in its own right; nor is `('T => Unit is Adj)[,]` a subtype of `('T => Unit)[,]`.
+
+## Impact on Existing Mechanisms
+
+This proposal would not modify or invalidate existing functionality (e.g.: `Int[][]` will continue to be a valid type in Q#), and thus is not a breaking change.
+
+If, in a future proposal, we were to unify multidimensional array with existing functionality using features outlined in _Anticipated Interactions with Future Modifications_, a breaking change would likely be required at that point in time.
```

Comment:
Can you be more specific here and elaborate? What do you mean by unifying multi-dimensional array functionality with existing one when introducing future features? I don't see a breaking change that would be necessary, and we should make sure that none is needed before we go ahead with this proposal.

Reply:
If I recall, this points to library functionality, as there's a lot of type and rank suffixes needed in the library proposal (e.g.: `Dot22C` for multiplying two two-dimensional arrays of complex numbers). Will delete here, in any case.